common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java

```diff
@@ -230,18 +261,31 @@ public T get(Object key) {
 
     public void put(T value) throws Exception {
       data.put(asKey(naturalKey.get(value)), value);
+      if (!naturalParentIndexName.isEmpty()) {
+        Comparable<Object> parentKey = asKey(getIndexAccessor(naturalParentIndexName).get(value));
+        NaturalKeys children =
+          parentToChildrenMap.computeIfAbsent(parentKey, k -> new NaturalKeys());
+        children.put(asKey(naturalKey.get(value)), true);
+      }
     }
 
     public void delete(Object key) {
       data.remove(asKey(key));
+      if (!naturalParentIndexName.isEmpty()) {
+        for (NaturalKeys v : parentToChildrenMap.values()) {
+          if (v.remove(asKey(key))) {
+            break;
+          }
+        }
```

Comment:
When a parent key in `parentToChildrenMap` points to empty `NaturalKeys`, we can also remove it?

Reply:
Yes, nothing will change if the NaturalKeys `v` doesn't contain `key`.

Reply:
Oh, I meant after `v.remove(asKey(key))`, if `v` is empty, can  we remove the  (parent key, empty NaturalKeys) from `parentToChildrenMap`?

Reply:
Well,  `parentToChildrenMap` is a concurrent map and checking emptiness costs time. 
The method here is to delete one entry. I think we can make it simple and keep it this way.